### PR TITLE
fix: send readonly command

### PR DESCRIPTION
**Pull request by proost**
I missed to send "READONLY" command for replica connection.

I want to integrate "READONLY" command with pipeline constructor, but it quite hard. has any idea?

### CODE REVIEW

**Comment by rueian** [comment]
Hi @proost, I think we should reuse this block: https://github.com/redis/rueidis/blob/65f705a8f7d7feab5d39dbd6725081f471aa1e6d/pipe.go#L273-L275

One more thing needs to be considered: A replica can be promoted to primary, meaning we should reset the 'READONLY' state with the `READWRITE` command.

**Comment by proost** [comment]
@rueian 
yes, you are right. replica can be promoted to primary. but it is hard to notice that replica is promoted. 
I thinkg that adding `ReplicaConnectionReadOnly` to `rueidis.ClientOption` and sending "READONLY" command in `_refresh` can be better. because use know when replica connection can be read-only connection. how do you think change like this?

**Comment by rueian** [comment]
Hi @proost,

I don’t think we need a new option. How about we modify the ReplicaOnly to true only for replica connections from the refresh function? And if we find a replica connection is promoted when refreshing, then we open a new one without ReplicaOnly.

**Comment by proost** [comment]
@rueian 

I change it. is okay to use `ReplicaOnly`? because it is breaking change.

File: cluster.go [hunk 1]
```diff
@@ -222,10 +217,42 @@ func (c *clusterClient) _refresh() (err error) {
...
+				timeout := c.opt.Dialer.Timeout
+				if timeout <= 0 {
+					timeout = DefaultDialTimeout
+				}
+
+				ctx, cancel := context.WithTimeout(context.Background(), timeout)
+				defer cancel()
+
+				cc.Do(ctx, cmds.NewCompleted([]string{"READONLY"})) // ignore error
+			}(cc)
```
Comment (rueian) [inline cluster.go:250]: Hi @proost,

Thank you for the changes, but this goroutine looks wired to me and I don't expect there will be breaking change by reusing the `ReplicaOnly`. I am afraid we have some misunderstanding here.

What I previously suggested is that the `pipe` will send `READONLY` command already if its `ReplicaOnly` is true.
So I think we can reuse this behavior by modifying the `c.connFn(addr, c.opt)` in the refresh function, and then the above goroutine is not necessary.

Comment (rueian) [reply cluster.go:250]: Given that the signature of the `connFn` is `func(dst string, opt *ClientOption) conn` which accepts a ClientOption pointer, I think we can copy the `c.opt` to `c.optReplica` with the `ReplicaOnly` set to true and then we can use `c.connFn(addr, c.optReplica)` for replica connections.

Comment (proost) [reply cluster.go]: [98040d7](https://github.com/redis/rueidis/pull/430/commits/98040d775392220820aea58a2738889c1607ad53)

@rueian 
I'm sorry about serious misunderstanding. I figured out.

File: cluster.go [hunk 2]
```diff
@@ -225,7 +220,15 @@ func (c *clusterClient) _refresh() (err error) {
 	for master, g := range groups {
 		conns[master] = connrole{conn: c.connFn(master, c.opt), replica: false}
 		for _, addr := range g.nodes[1:] {
-			conns[addr] = connrole{conn: c.connFn(addr, c.opt), replica: true}
+			var cc conn
+			if c.opt.ReplicaOnly {
+				opt := *c.opt
```
Comment (rueian) [inline cluster.go:225]: I am still a little confused. So now, are we requiring users to set `ReplicaOnly` as well even if they use `SendToReplicas`?

I thought we would only set `ReplicaOnly` transparently for users here to the `opt`.

Comment (proost) [reply cluster.go]: fix it. [3ec3548](https://github.com/redis/rueidis/pull/430/commits/3ec3548e8207c497029f37acad869db5e0d3e5f5) 

when `SendToReplicas` is configured and `ReplicaOnly` is true,
- connection to master: do not send "READONLY"
- connection to replica: send "READONLY"

when `SendToReplicas` is configured and `ReplicaOnly` is false,
- connection to master: do not send "READONLY"
- connection to replica: do not send "READONLY"

when `SendToReplicas` is not configured and `ReplicaOnly` is true,
- connection to master: send "READONLY"

Comment (rueian) [reply cluster.go]: Why we don't send "READONLY" to replicas when SendToReplicas is configured and ReplicaOnly is false?

Isn't it the original issue you want to fix?

File: cluster.go [hunk 3]
```diff
@@ -223,7 +218,11 @@ func (c *clusterClient) _refresh() (err error) {
 	groups := result.parse(c.opt.TLSConfig != nil)
 	conns := make(map[string]connrole, len(groups))
 	for master, g := range groups {
-		conns[master] = connrole{conn: c.connFn(master, c.opt), replica: false}
+		opt := *c.opt
+		if c.opt.SendToReplicas != nil {
+			opt.ReplicaOnly = false
+		}
+		conns[master] = connrole{conn: c.connFn(master, &opt), replica: false}
```
Comment (rueian) [inline cluster.go:225]: The current code above will allocate a copy of `c.opt` on the heap every time.
We should do it better by preparing the copied `opt` once in advance.

File: cluster.go (hunk 2 above)
Comment (proost) [reply cluster.go]: [cd6a674](https://github.com/redis/rueidis/pull/430/commits/cd6a674ec54ce392de2ff6e0d268fc6cb877e344)

Yes, it doesn't matter HOW. i just want to send "READONLY" command to replica connections automatically. but if you think read-only mode as a default is better than i changed it.

File: cluster.go (hunk 3 above)
Comment (proost) [reply cluster.go]: [cd6a674](https://github.com/redis/rueidis/pull/430/commits/cd6a674ec54ce392de2ff6e0d268fc6cb877e344)

yes. you're right.

File: cluster.go [hunk 4]
```diff
@@ -222,10 +217,18 @@ func (c *clusterClient) _refresh() (err error) {
...
 	groups := result.parse(c.opt.TLSConfig != nil)
 	conns := make(map[string]connrole, len(groups))
+	masterOpt := *c.opt
+	if c.opt.SendToReplicas != nil {
+		masterOpt.ReplicaOnly = false
+	}
+	replicaOpt := *c.opt
+	if c.opt.SendToReplicas != nil {
+		replicaOpt.ReplicaOnly = !c.opt.ReplicaOnly
+	}
```
Comment (rueian) [inline cluster.go:227]: It is better now. But it still allocates `masterOpt` and `replicaOpt` in every refresh.
We can make it even better by storing a `c.replicaOpt` in the `newClusterClient`.

Comment (rueian) [reply cluster.go:227]: >  i just want to send "READONLY" command to replica connections automatically.

Agreed. So, how about this:

when `SendToReplicas` is configured and `ReplicaOnly` is true,

* return `ErrReplicaOnlyConflict`

when SendToReplicas is configured and ReplicaOnly is false,

* connection to master: use `c.connFn(addr, c.opt)`
* connection to replica: use `c.connFn(addr, c.replicaOpt)`

when SendToReplicas is not configured and ReplicaOnly is true,

* connection to all nodes: use `c.connFn(addr, c.opt)`, we don't need `c.replicaOpt` in this case.

Comment (proost) [reply cluster.go]: [5702ce3](https://github.com/redis/rueidis/pull/430/commits/5702ce3e8deb82dd5848d1adba36b07842e94e57)

Good! i changed it.

File: cluster.go [hunk 5]
```diff
@@ -241,7 +254,8 @@ func (c *clusterClient) _refresh() (err error) {
 
 	c.mu.RLock()
 	for addr, cc := range c.conns {
-		if fresh, ok := conns[addr]; ok {
+		fresh, ok := conns[addr]
+		if ok && cc.replica == fresh.replica {
```
Comment (rueian) [inline cluster.go:258]:
```suggestion
		if ok && (cc.replica == fresh.replica || c.opt.SendToReplicas == nil) {
```

Comment (rueian) [reply cluster.go:258]: We can still reuse the conn if c.opt.SendToReplicas is not configured.

Comment (proost) [reply cluster.go]: [87e2b39](https://github.com/redis/rueidis/pull/430/commits/87e2b39036c247c1106d95207277848d3c7b1f2c)

Comment (rueian) [reply cluster.go]: Hi @proost, thank you very much! It looks great!

**Pull request merged**